### ISSUE TITLE
How to manually add annotation by hand after each epoch?

### ISSUE THREAD

**Issue by whaowhao** [ElementAI/baal#160]
Assuming baal returns a list of datums to be labeled by humans after each epoch, how do I connect it to Amazon Mechanical Turk to get them actually labeled for the next epoch's training? Can the training process be stalled while humans label these examples?

**Comment by Dref360** [comment]
Hello,

I'll rework this tutorial, but we do have a tutorial on [how to use BaaL for production](https://baal.readthedocs.io/en/latest/notebooks/baal_prod_cls.html). 

I'm not knowledgeable in Amazon Turk, but I think you could have a loop similar to the script below. In this case, `get_label_from_amazon_turk` would block until we have the label.

Let me know how we can help! We would gladly improve our API to better suit your use case.


```python
baal_model : ModelWrapper = ...
active_learning_ds: ActiveLearningDataset = ...
heuristic : BALD = ...
QUERY_SIZE = 10 # Amount of label to get between retraining.
NUM_STEPS = 100 # How many round of active learning.

for step in range(NUM_STEPS):
  # 2. Train the model for a few epoch on the training set.
  print(f"Training on {len(active_learning_ds)} items!")
  baal_model.train_on_dataset(active_learning_ds, optimizer, batch_size=16, epoch=5, use_cuda=USE_CUDA)
  baal_model.test_on_dataset(test_dataset, batch_size=16, use_cuda=USE_CUDA)

  print("Metrics:", {k:v.avg for k,v in baal_model.metrics.items()})

  # 3. Select the K-top uncertain samples according to the heuristic.
  pool = active_learning_ds.pool
  if len(pool) == 0:
    print("No items to label.")
    break

  predictions = baal_model.predict_on_dataset(pool, batch_size=16, iterations=15, use_cuda=USE_CUDA, verbose=False)
  top_uncertainty = heuristic(predictions)[:QUERY_SIZE]
  # Translate the indices back to the full dataset.
  raw_top_uncertainty = active_learning_ds._pool_to_oracle_index(top_uncertainty)
  # 4. Label those samples.
  labels = get_label_from_amazon_turk([train_dataset.files[idx] for idx in raw_top_uncertainty])
  active_learning_ds.label(top_uncertainty, labels)
```

**Comment by noknok00** [comment]
Hi @Dref360 , I am working on a similar (school) project, where I have a front-end for Image Labelling (kind-of Label Studio) and I am thinking on using BAAL for the Active Learning in the back-end.
The solution I have in mind is to, rather than having a loop (NUM_STEPS), I would convert it to a function (the loop) to be called every time the Annotator (from the front-end) returns the labelled images.
The difference in my function (compared to the current loop) is that I would load the model (from the saved checkpoint) to load the new labelled data and save the new checkpoint at the end of this function (after sending the new unlabelled data to the Annotator's queue).

I do have three questions, I am completely new on Machine Learning and specifically on Active Learning.
1. since I am saving the current state of the model (with torch.save), does it affect in some way to resume the training on a different moment? (equivalent to run the loop once, and then re-loading the model to re-run the loop once again, compared to run the loop continuously as in the example provided).
2. I have been reading your documentation (and torch lightning docs, and lightning flash docs) and have a hard time understanding how to actually label manually, like all the tutorial focus on a research use-case, but don't give any "hints" on how to feed manual labels to the codes given as example. Maybe I am missing something. The example "how to use BaaL for production." is the only reason I understood how to use a manual labelling function.

continuing on the second point.
I wanted to use lighting flash in my project (to use with BAAL), but I have spent the whole day digging into the source code to understand have can I implement a Manual Label (Oracle) function. in Lightning Flash, the BAAL framework is so much wrapped up that I don't think is easy to add an Oracle function.

I am talking about this tutorial:
https://lightning-flash.readthedocs.io/en/latest/integrations/baal.html

`# 3.2 Create the active learning loop and connect it to the trainer
active_learning_loop = ActiveLearningLoop(label_epoch_frequency=1)
active_learning_loop.connect(trainer.fit_loop)
trainer.fit_loop = active_learning_loop`

Can you point me in the right direction on how may I insert an Oracle function (to manually label data) by using Lightning Flash with BAAL?

3. my third question is, it is intentional that "ActiveLearningLoop" class do not have a flag (or an easy way) to allow manual annotation?
in the "step" method, you could add a flag to allow the user to manually annotate the data.

`                if len(to_label) > 0:
                    self.dataset.label(to_label[: self.query_size])
                    return True
`

or is this class just meant for demonstration?


Thanks and sorry for the long text.

**Comment by Dref360** [comment]
Hi!

Yeah the Lightning Flash integration is mostly for research right now, but you could implement your own ActiveDataModule and change the `label` function [here](https://github.com/PyTorchLightning/lightning-flash/blob/5dd695f419703de65bc06ab4983ef5ce891669e9/flash/image/classification/integrations/baal/data.py#L159). I think the Flash team is working with Label Studio right now to get the "production" side going. We are currently not involved in this part of the project.

To answer your questions,

1. It shouldn't pose an issue as we keep track of what is labelled in ActiveLearningDataset.state_dict. You might need to save the actual labels yourself thought.
2. I think "How to use BaaL for Prod" is a good way to start. You can use Pytorch Lightning to train, evaluate and get predictions. You then have your `labelling function` that ask for labels for the most uncertain examples.
3. ActiveLearningLoop is made for research currently. We could add methods to support your usecase. I'm thinking something like `ActiveLearningLoop.get_indices_to_label` that does not call `label`, but simply returns the indices. Would that work?
We are really open to modifying our API to support your usecase!

Don't hesitate to join our Slack for easier communication.

**Comment by parmidaatg** [comment]
@Dref360 can we close this issue safely?

**Comment by Dref360** [comment]
Yes
